### PR TITLE
Fix kinsol

### LINKED ISSUE #12223: KINSOL not working
#12216 introduced a bug that only shows on some versions of sundials. 

I'm working on it now.

### PR DESCRIPTION
**Pull request by luca-heltai**
This should fix all remaining issues with KINSOL.

closes #12223.

### CODE REVIEW

**Comment by luca-heltai** [comment]
@sebproell , this addresses everything we discussed on slack. Thanks to you and to @nicola-giuliani  for the help.

**Review by peterrum — APPROVED** [review]

File: source/sundials/kinsol.cc [hunk 1]
```diff
@@ -354,17 +373,26 @@ namespace SUNDIALS
...
-    AssertKINSOL(status);
 
     status = KINSetUserData(kinsol_mem, static_cast<void *>(this));
     AssertKINSOL(status);
 
+    // This must be called before KINSetMAA
     status = KINSetNumMaxIters(kinsol_mem, data.maximum_non_linear_iterations);
     AssertKINSOL(status);
 
+    // From the manual: this must be called BEFORE KINInit
```
Comment (bangerth) [inline source/sundials/kinsol.cc:386]: how bizarre...

File: source/sundials/kinsol.cc [hunk 2]
```diff
@@ -524,6 +542,8 @@ namespace SUNDIALS
     reinit_vector = [](VectorType &) {
       AssertThrow(false, ExcFunctionNotProvided("reinit_vector"));
     };
+
+    setup_jacobian = [](const VectorType &, const VectorType &) { return 0; };
```
Comment (bangerth) [inline source/sundials/kinsol.cc:546]: Wouldn't you want this function to throw an assert, just like 3 lines higher up?

**Review by bangerth — COMMENTED** [review]
I have just one question, but the rest works for me.

I didn't see where the original bug was. To me it looks like you are only shuffling code into the right place, and support a couple of things that may previously not have been supported (like the `get_constraints()` functionality). What was the bug and which part fixes it?

**Comment by bangerth** [comment]
The CI check shows this failure:
```

[ 33%] Generating kinsol_02.debug/interrupt_guard.cc
Scanning dependencies of target sundials.kinsol_02.debug
[ 66%] Building CXX object CMakeFiles/sundials.kinsol_02.debug.dir/kinsol_02.cc.o
[ 50%] Building CXX object CMakeFiles/sundials.kinsol_02.debug.dir/kinsol_02.debug/interrupt_guard.cc.o
[ 83%] Linking CXX executable kinsol_02.debug/kinsol_02.debug
[166%] Built target sundials.kinsol_02.debug
Scanning dependencies of target sundials.kinsol_02.debug.diff
[216%] Generating kinsol_02.debug/output
sundials/kinsol_02.debug: BUILD successful.
sundials/kinsol_02.debug: RUN failed. ------ Return code 134
sundials/kinsol_02.debug: RUN failed. ------ Result: /home/dealii/build/tests/sundials/kinsol_02.debug/failing_output
sundials/kinsol_02.debug: RUN failed. ------ Partial output:

sundials/kinsol_02.debug: RUN failed. ------ Additional output on stdout/stderr:

terminate called after throwing an instance of 'std::bad_function_call'
  what():  bad_function_call
```
Can you take another look?

File: source/sundials/kinsol.cc (hunk 1 above)
Comment (nicola-giuliani) [reply source/sundials/kinsol.cc:386]: @bangerth this was the bug. @sebproell discovered that if this function is called after the initialisation a vector of N_Vector is left empty so the fixed point iterations end up segfaulting. Then we looked carefully at the manual and discovered the fix

**Comment by luca-heltai** [comment]
I'll be looking at the failure later today. It is happening in the old version of sundials, not the new one.

File: source/sundials/kinsol.cc (hunk 2 above)
Comment (luca-heltai) [reply source/sundials/kinsol.cc:546]: Not really. We claim it is optional. Indeed the older version worked without this method. The new version, however, requires the setup. Providing a "do-nothing" setup by default should maintain backward compatibility and allow one not to specify any setup if required.

Comment (bangerth) [reply source/sundials/kinsol.cc:541]: I see. In that case, can you please move this line out of this function? The function is called `set_functions_to_trigger_an_assert()` :-)

I think the place where we do the general KINSOL set-up would be right.

**Comment by bangerth** [comment]
Looks good to me, and it also tests correctly, but I would still like it if you could address this one comment: https://github.com/dealii/dealii/pull/12254/files#r636201155

**Comment by bangerth** [comment]
OK to merge with this one change.

**Comment by luca-heltai** [comment]
see my comment in #12200. Should we rename the function in this class, or the ones in IDA?

**Review by bangerth — APPROVED** [review]

**Comment by bangerth** [comment]
Let's get this one merged as-is and leave the renaming for later.

**Comment by bangerth** [comment]
later = later today, in a separate patch

**Pull request merged**